test/util/malicious/hasher.go

```diff
@@ -0,0 +1,225 @@
+package malicious
+
+import (
+	"bytes"
+	"errors"
+	"hash"
+
+	"github.com/celestiaorg/nmt"
+	"github.com/celestiaorg/nmt/namespace"
+)
+
+// NOTE: This file is a copy of the original nmt.Hasher implementation, but with
+// all of the validation logic removed. This allows for malicious nodes to create nmt roots that are out of order.
+
+const (
+	LeafPrefix = 0
+	NodePrefix = 1
+)
+
+var _ hash.Hash = (*Hasher)(nil)
+
+var (
+	ErrUnorderedSiblings         = errors.New("NMT sibling nodes should be ordered lexicographically by namespace IDs")
+	ErrInvalidNodeLen            = errors.New("invalid NMT node size")
+	ErrInvalidLeafLen            = errors.New("invalid NMT leaf size")
+	ErrInvalidNodeNamespaceOrder = errors.New("invalid NMT node namespace order")
+)
+
+type Hasher struct {
+	baseHasher   hash.Hash
+	NamespaceLen namespace.IDSize
+
+	// The "ignoreMaxNs" flag influences the calculation of the namespace ID
+	// range for intermediate nodes in the tree i.e., HashNode method. This flag
+	// signals that, when determining the upper limit of the namespace ID range
+	// for a tree node, the maximum possible namespace ID (equivalent to
+	// "NamespaceLen" bytes of 0xFF, or 2^NamespaceLen-1) should be omitted if
+	// feasible. For a more in-depth understanding of this field, refer to the
+	// "HashNode".
+	ignoreMaxNs      bool
+	precomputedMaxNs namespace.ID
+
+	tp   byte   // keeps type of NMT node to be hashed
+	data []byte // written data of the NMT node
+}
+
+func (n *Hasher) IsMaxNamespaceIDIgnored() bool {
+	return n.ignoreMaxNs
+}
+
+func (n *Hasher) NamespaceSize() namespace.IDSize {
+	return n.NamespaceLen
+}
+
+// NewHasher returns a nmt.Hasher that
+func NewBlindHasher(baseHasher hash.Hash, nidLen namespace.IDSize, ignoreMaxNamespace bool) *Hasher {
+	return &Hasher{
+		baseHasher:       baseHasher,
+		NamespaceLen:     nidLen,
+		ignoreMaxNs:      ignoreMaxNamespace,
+		precomputedMaxNs: bytes.Repeat([]byte{0xFF}, int(nidLen)),
+	}
+}
+
+// Size returns the number of bytes Sum will return.
+func (n *Hasher) Size() int {
+	return n.baseHasher.Size() + int(n.NamespaceLen)*2
+}
+
+// Write writes the namespaced data to be hashed.
+//
+// Requires data of fixed size to match leaf or inner NMT nodes. Only a single
+// write is allowed.
+// It panics if more than one single write is attempted.
+// If the data does not match the format of an NMT non-leaf node or leaf node, an error will be returned.
```

Comment:
[no change needed] maintainers should consider enabling a linter that auto wraps comment length